main.go

```diff
@@ -178,57 +144,59 @@ func main() {
 		})
 	}
 	{
+		l := log.With(l, "component", "writer")
 		t := time.NewTicker(opts.Period)
 		ctx, cancel := context.WithCancel(ctx)
 
 		g.Add(func() error {
-			level.Info(logger).Log("msg", "starting the remote-write client")
+			level.Info(l).Log("msg", "starting the remote-write client")
 			for {
 				select {
 				case <-t.C:
-					if err := write(ctx, endpointWrite, opts.Token, generate(opts.Labels)); err != nil {
-						level.Error(logger).Log("msg", "failed to make request", "err", err)
+					if err := write(ctx, opts.WriteEndpoint, opts.Token, generate(opts.Labels), l); err != nil {
+						m.remoteWriteRequests.WithLabelValues("error").Inc()
+						level.Error(l).Log("msg", "failed to make request", "err", err)
+					} else {
+						m.remoteWriteRequests.WithLabelValues("success").Inc()
 					}
 				case <-ctx.Done():
-					return nil
+					return reportResults(l, m.remoteWriteRequests, opts.SuccessThreshold)
 				}
 			}
 		}, func(_ error) {
 			t.Stop()
 			cancel()
 		})
 	}
-	{
-		var cancel func()
 
-		if opts.Duration != 0 {
-			ctx, cancel = context.WithTimeout(ctx, opts.Duration)
-		} else {
-			ctx, cancel = context.WithCancel(ctx)
-		}
+	if opts.ReadAddress != nil {
+		{
+			l := log.With(l, "component", "querier")
+			t := time.NewTicker(opts.Period)
 
-		t := time.NewTicker(opts.Period)
+			var cancel func()
+			if opts.Duration != 0 {
+				ctx, cancel = context.WithTimeout(ctx, opts.Duration)
+			} else {
+				ctx, cancel = context.WithCancel(ctx)
+			}
 
-		if opts.EndpointRead != "" {
 			g.Add(func() error {
-				level.Info(logger).Log("msg", "start querying for metrics")
+				level.Info(l).Log("msg", "waiting for initial delay before querying for metrics")
+				time.Sleep(opts.InitialQueryDelay)
```

Comment:
sleeping like this makes the routine un-interruptible, so if someone set the initial delay to 30 seconds and tried to SIGINT immediately after starting, they wouldn't be able to and would have to wait 30s. To make this interruptible, we could do:
```go
select {
case ctx.Done():
	return nil
case <-time.After(opts.InitialQueryDelay):
}
```